File: src/main/java/no/nav/personopplysninger/features/norg2/Norg2Consumer.java
package no.nav.personopplysninger.features.norg2;

import no.nav.log.MDCConstants;
import no.nav.personopplysninger.features.ConsumerFactory;
import no.nav.personopplysninger.features.norg2.domain.Norg2Enhet;
import no.nav.personopplysninger.features.norg2.domain.Norg2EnhetKontaktinfo;
import no.nav.personopplysninger.features.personalia.exceptions.ConsumerException;
import no.nav.tps.person.Personinfo;
import org.slf4j.MDC;

import javax.ws.rs.client.Client;
import javax.ws.rs.client.Invocation;
import javax.ws.rs.core.Response;
import java.net.URI;

import static javax.ws.rs.core.Response.Status.Family.SUCCESSFUL;

public class Norg2Consumer {
    
    private Client client;
    private URI endpoint;

    public Norg2Consumer(Client client, URI endpoint) {
        this.client = client;
        this.endpoint = endpoint;
    }

    public Norg2Enhet hentEnhet(String geografisk) {
        Invocation.Builder request = buildEnhetRequest(geografisk, "enhet/navkontor");
        return hentEnhet(request);
    }

    public Norg2EnhetKontaktinfo hentKontaktinfo(String enhetsnr) {
        Invocation.Builder request = buildKontaktinfoRequest(enhetsnr, "enhet/kontaktinformasjon");
        return hentKontaktinfo(request);
    }

    private Invocation.Builder buildEnhetRequest(String geografisk, String path) {
        return client.target(endpoint)
                .path(path + "/" + geografisk)
                .request()
                .header("Nav-Call-Id", MDC.get(MDCConstants.MDC_CALL_ID))
                .header("Nav-Consumer-Id", ConsumerFactory.CONSUMER_ID);

    }

    private Invocation.Builder buildKontaktinfoRequest(String enhetsnr, String path) {
        return client.target(endpoint)
                .path(path)
                .request()
                .header("Nav-Call-Id", MDC.get(MDCConstants.MDC_CALL_ID))
                .header("Nav-Consumer-Id", ConsumerFactory.CONSUMER_ID)
                .header("enhetsnr", enhetsnr);
    }

    private Norg2Enhet hentEnhet(Invocation.Builder request) {
        try (Response response = request.get()) {
            return readEnhetResponse(response);
        } catch (ConsumerException e) {
            throw e;
        } catch (Exception e) {
            String msg = "Forsøkte å konsumere REST-tjenesten TPS-proxy. endpoint=[" + endpoint + "].";
            throw new ConsumerException(msg, e);
        }
    }

    private Norg2EnhetKontaktinfo hentKontaktinfo(Invocation.Builder request) {
        try (Response response = request.get()) {
            return readKontaktinfoResponse(response);
        } catch (ConsumerException e) {
            throw e;
        } catch (Exception e) {
            String msg = "Forsøkte å konsumere REST-tjenesten TPS-proxy. endpoint=[" + endpoint + "].";
            throw new ConsumerException(msg, e);
        }
    }


    private Norg2Enhet readEnhetResponse(Response r) {
        if (!SUCCESSFUL.equals(r.getStatusInfo().getFamily())) {
            String msg = "Forsøkte å konsumere REST-tjenesten TPS-proxy. endpoint=[" + endpoint + "], HTTP response status=[" + r.getStatus() + "].";
            throw new ConsumerException(msg + " - " + ConsumerFactory.readEntity(String.class, r));
        } else {
            return ConsumerFactory.readEntity(Norg2Enhet.class, r);
        }
    }

    private Norg2EnhetKontaktinfo readKontaktinfoResponse(Response r) {
        if (!SUCCESSFUL.equals(r.getStatusInfo().getFamily())) {
            String msg = "Forsøkte å konsumere REST-tjenesten TPS-proxy. endpoint=[" + endpoint + "], HTTP response status=[" + r.getStatus() + "].";
            throw new ConsumerException(msg + " - " + ConsumerFactory.readEntity(String.class, r));
        } else {
            return ConsumerFactory.readEntity(Norg2EnhetKontaktinfo.class, r);
        }
    }

}

File: src/main/kotlin/no/nav/personopplysninger/oppslag/norg2/Norg2Consumer.kt
package no.nav.personopplysninger.oppslag.norg2

import com.fasterxml.jackson.module.kotlin.readValue
import no.nav.common.log.MDCConstants
import no.nav.personopplysninger.oppslag.norg2.domain.Norg2Enhet
import no.nav.personopplysninger.oppslag.norg2.domain.Norg2EnhetKontaktinfo
import no.nav.personopplysninger.util.CONSUMER_ID
import no.nav.personopplysninger.util.ConsumerException
import no.nav.personopplysninger.util.JsonDeserialize.objectMapper
import no.nav.personopplysninger.util.consumerErrorMessage
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.slf4j.MDC
import java.net.URI
import javax.ws.rs.client.Client
import javax.ws.rs.client.Invocation
import javax.ws.rs.core.Response.Status.Family.SUCCESSFUL

class Norg2Consumer(private val client: Client, private val endpoint: URI) {

    private var logger: Logger = LoggerFactory.getLogger(javaClass)

    fun hentEnhetDersomGyldig(geografisk: String): Norg2Enhet? {
        return try {
            hentEnhet(geografisk)
        } catch (e: ConsumerException) {
            logger.warn("Feil oppstod ved henting av enhet: ", e)
            null
        }
    }

    private fun hentEnhet(geografisk: String): Norg2Enhet? {
        val request = buildEnhetRequest(geografisk)
        return request.readResponse()
    }

    fun hentKontaktinfo(enhetsnr: String): Norg2EnhetKontaktinfo {
        val request = buildKontaktinfoRequest(enhetsnr)
        return request.readResponse()
    }

    private fun buildEnhetRequest(geografisk: String): Invocation.Builder {
        return client.target(endpoint)
            .path("enhet/navkontor/$geografisk")
            .request()
            .header("Nav-Call-Id", MDC.get(MDCConstants.MDC_CALL_ID))
            .header("Nav-Consumer-Id", CONSUMER_ID)
    }

    private fun buildKontaktinfoRequest(enhetsnr: String): Invocation.Builder {
        return client.target(endpoint)
            .path("enhet/$enhetsnr/kontaktinformasjon")
            .request()
            .header("Nav-Call-Id", MDC.get(MDCConstants.MDC_CALL_ID))
            .header("Nav-Consumer-Id", CONSUMER_ID)
            .header("enhetsnr", enhetsnr)
    }

    private inline fun <reified T> Invocation.Builder.readResponse(): T {
        get().use { response ->
            val responseBody = response.readEntity(String::class.java)
            if (SUCCESSFUL != response.statusInfo.family) {
                throw ConsumerException(consumerErrorMessage(endpoint, response.status, responseBody))
            }
            return objectMapper.readValue(responseBody)
        }
    }
}
